pkg/controllermanager/controller/shoot/maintenance/reconciler.go

```diff
@@ -199,31 +207,65 @@ func (r *Reconciler) reconcile(ctx context.Context, log logr.Logger, shoot *gard
 			v = workerPoolSemver.String()
 			maintainedShoot.Spec.Provider.Workers[i].Kubernetes.Version = &v
 			return nil
-		}, name)
+		})
 		if err != nil {
-			// continue execution to allow the machine image version update
-			workerLog.Error(err, "Could not maintain kubernetes version for worker pool")
+			// continue execution to allow other maintenance activities to continue
+			workerLog.Error(err, "Could not maintain Kubernetes version for worker pool")
+		}
+
+		if reasonForUpdate != nil {
+			workerToReasonForKubernetesUpdate[pool.Name] = *reasonForUpdate
+		}
+
+		if successMsg != nil {
+			workerToSuccessfulKubernetesUpdate[pool.Name] = *successMsg
 		}
 
-		reasonForKubernetesUpdate = append(reasonForKubernetesUpdate, reasonForWorkerPoolKubernetesUpdate...)
+		if failureMsg != nil {
+			workerToFailedKubernetesUpdate[pool.Name] = *failureMsg
+		}
 	}
-	operations = append(operations, reasonForKubernetesUpdate...)
 
 	operation := maintainOperation(maintainedShoot)
 	if operation != "" {
 		operations = append(operations, fmt.Sprintf("Added %q operation annotation", operation))
 	}
 
-	if len(operations) > 0 {
+	requirePatch := len(operations) > 0 || reasonControlPlaneKubernetesUpdate != nil || len(workerToReasonForKubernetesUpdate) > 0 || len(workerToReasonForImageUpdate) > 0
+	if requirePatch {
 		patch := client.MergeFrom(shoot.DeepCopy())
+
+		// make sure to both include successful and failed maintenance operations
+		description, failureReason := buildMaintenanceMessages(
+			reasonControlPlaneKubernetesUpdate,
+			successMsgControlPlaneKubernetesUpdate,
+			failureMsgControlPlaneKubernetesUpdate,
+			workerToReasonForKubernetesUpdate,
+			workerToSuccessfulKubernetesUpdate,
+			workerToFailedKubernetesUpdate,
+			workerToReasonForImageUpdate,
+			workerToSuccessfulImageUpdate,
+			workerToFailedImageUpdate)
+
+		// append also other maintenance operation
+		if len(operations) > 0 {
+			description = fmt.Sprintf("%s, %s", description, strings.Join(operations, ", "))
+		}
+
 		shoot.Status.LastMaintenance = &gardencorev1beta1.LastMaintenance{
-			Description:   strings.Join(operations, ", "),
+			Description:   description,
 			TriggeredTime: metav1.Time{Time: r.Clock.Now()},
-			State:         gardencorev1beta1.LastOperationStateProcessing,
+			State:         gardencorev1beta1.LastOperationStateSucceeded,
 		}
 
-		// First dry run the update call to check if it can be executed successfully.
-		// If not shoot maintenance is marked as failed and is retried only in
+		// if any maintenance operation failed, set the status to 'Failed' and retry in the next maintenance cycle
+		if failureReason != "" {
+			shoot.Status.LastMaintenance.State = gardencorev1beta1.LastOperationStateFailed
+			shoot.Status.LastMaintenance.FailureReason = &failureReason
+		}
```

Comment:
I understand your motivation here that you want to put the list of errors here and make the state fail, but this is not correct.
The issue is if failurerReasons is not nil and there are some changes that can be applied and that don't fail during the dry run phase in that case the final state of the last maintenance will be failed, but some changes would have been applied, this doesn't seem correct to me as state failed give the sense that nothing would have done during maintenance time.
I would rather prefer to make the state succeed and in the description, you can put successful operations and failed operations in groups describing what changes were successful and what were not.

Reply:
> Overall, the GCM currently assumes that its strategies to determine a higher version and any other internal logic never fail. That's why the maintenance condition on the Shoot can only fail if the API server update of the Shoot is rejected. I don't think that is valid, as it hides possible errors (maintenance activities can fail such as when its not possible to find a suitable higher version due to CloudProfile misconfiguration or because the machine image is EOL + there might be more maintenance operations with failure possibilities in the future).

IMO, why should user care if the operator has misconfigured the cloud profile? User should ideally not see this error.

Reply:
Also consider the below scenario in the last maintenance - 
`description` - Kubernetes version is upgraded to 1.25. 
`failureReason` - Machine image version update failed because some error in cloudprofile config.
`state` - failed.

From the above one will think(or at least I get the same sense) that the last maintenance operation was failed but in actual the Kubernetes version is upgraded.

Reply:
> From the above one will think(or at least I get the same sense) that the last maintenance operation was failed but in actual the Kubernetes version is upgraded.

That's a good point. The Description should not include a forgone conclusion, but should probably be formulated neutral "Update Kubernetes version to 1.25"

Reply:
> IMO, why should user care if the operator has misconfigured the cloud profile? User should ideally not see this error.

Isn't that how most of our controllers work? If there is an error, being a configuration error or an internal error during the reconciliation that prevents a successful reconciliation, we would indicate that in the resource status?

@acumino maybe you and the other maintainers can make a decision and I'll update the PR accordingly.

Reply:
@acumino @ialidzhikov @oliver-goetz @rfranzke I would appreciate a decision so we can move forward, thank you!

Reply:
I don't feel that I have all of the context but If I managed to follow correctly, the question is what to be the status when there is one maintenance operation that succeeded and other that failed:

```yaml
status:
  lastMaintenance:
    description: "Kubernetes version is upgraded to 1.25."
    failureReason: "Machine image version update failed."
    state: Error or Succeeed?
```

I would vote for `state=Error` for error visibility reasons - if I am an ops person I would expect to see `state=Error` in the above case and not `state=Succeeded` and the actual error set in the `failureReason`. It also looks strange to have a `failureReason` set when state is `Succeeded`. There is then the question why some operation was performed when another operation failed.

Currently most of our controllers work in the following pattern - if they encounter error, they fail right away and do not try to continue. But as far as I see, we ignore some errors in the maintenance controller and try to patch the shoot if we manage to update something. For example, the maintenance controller would try to update the K8s version even if the machine image version update failed. I don't know what is the exact reason for it but we wouldn't be discussing what should be the state (error or succeeded) if the maintenance controller was transactional - when there is err, it fails and does not update anything; when there is no err, it updates everything that needs update. Does someone recall why the maintenance controller was implemented in such non-transactional way?

I also don't understand why we also try to fix also this issue in this PR. IIUC, it is a separate issue that some maintenance controller failures are not visible in the `.status.lastMaintenance` field and only logged. Only the dry-run update failure would be reported in the in the `.status.lastMaintenance` field in the current master. So, it can be a dedicated issue that can be tackled out of this PR. This PR is anyways big enough and requires context to follow. Adding such not directly related changes to PR makes reviewer's live harder.

Reply:
Thanks for the review. In retrospect, it would have been better to not address this topic in the current PR knowing its apparent controversy. I thought it's an opportune moment, as more failure points (more strategies that can fail) are introduced.
I would have hoped that if the change is generally agreed to be valuable, to leave it in this PR due to economical reasons, as I would otherwise have to re-engineer and re-test large parts of the PR, only to propose the same thing later. 

If you want to open up the discussion to change the GCM to be transactional, then I surely agree that's outside the scope of this PR (and also won't be done by me).

Reply:
In a discussion with @danielfoehrKn and @acumino we agreed to:
- use `state: Failed` when there are some maintenance operations succeeded but other(s) - failed
  - Reason: as outlined above, make operator lives easy - show the failure in the lastMaintenance field; otherwise the failure reason would be present only in the GCM logs. The raised concern was that in most of the cases the raised error might be related due to misconfigured CloudProfile and the question was "Should we expose this to the end users when they cannot fix it themselves (misconfigured CloudProfile)?". The arguments were that we already expose a lot of things in the Shoot status and usually end users cannot help themselves to fix them (for example Shoot failing to be scheduled because there is no available Seed).
- build and display proper description on which operations succeeded and which operations failed in the `description` field to make it clear that some operations succeeded but others failed.
- populate `failureReason` with the operations that failed
- keep the controller non-transactional - it seems to be a desired behaviour to update version if possible, when other update fails. Otherwise, we could unnecessarily block a version upgrade because some other upgrade operation failed before it.